### PR TITLE
[FLOC-1480] Allow some builders to fail without reporting loudly.

### PR DESCRIPTION
**Pull request by tomprince**
https://clusterhq.atlassian.net/browse/FLOC-1480

This makes it so that there is a list of builder in `config.yml` that don't get reported to github, and failures on master don't get reported to zulip's engineering stream.

- there should be a way to force a build that will report status (so that branches that are supposed to fix things can have status reported to github).
- This currently keeps the list of builders in `config.yml`. It might make sense to have something that is easier to change. Although, perhaps it isn't worth the time it would take to implement.
- A different option is to look at all the historical builds, and not report builds that have never passed on master.

### CODE REVIEW

**Comment by adamtheturtle** [comment]
The issue specifies that this is in particular for new builders. It seems to me that this should be the other way around therefore - there is a configurable list of builders which *do* get statuses reported. This means that someone adding a builder has one final step to add to this list, instead of a step to add to this list, then later remove from this list.

**Comment by adamtheturtle** [comment]
So that I don't have to spin up a staging instance with this on, please can you link to somewhere where this is working, or perhaps some screenshots of the modified pages.

File: flocker_bb/zulip_status.py [hunk 1]
```diff
@@ -22,7 +22,10 @@
 }
 
 
-@attributes(['zulip', 'stream', 'critical_stream'])
+@attributes([
+    'zulip', 'stream', 'critical_stream',
+    Attribute('failing_builders', default_factory=frozenset),
```
Comment (adamtheturtle) [inline flocker_bb/zulip_status.py]: :question: Document at least the new attribute.

**Comment by adamtheturtle** [comment]
Thanks @tomprince this looks like it will be useful. Please let me know what you think of the proposal to swap the `failing_builders` attribute for a list of expected-passing builders. The current approach (or perhaps just the current variable names) creates confusion between the builders which are failing now, as expected perhaps because they are new / being changed, and those which are failing now because of other problems.

The proposals in the description of this PR seem reasonable, but if my understanding of JIRA is correct, then this is being worked on in this sprint as a blocker to other issues. Those proposals seem more like follow-ups than blockers to me.

> A different option is to look at all the historical builds, and not report builds that have never passed on master.

Personally, I'd rather have something explicit. Please continue if you agree with my proposal, or discuss if you don't.

**Comment by tomprince** [comment]
> The issue specifies that this is in particular for new builders. It seems to me that this should be the other way around therefore - there is a configurable list of builders which do get statuses reported. This means that someone adding a builder has one final step to add to this list, instead of a step to add to this list, then later remove from this list.

There are a couple of reasons I don't think this is a good idea:
- The steady state should be that there are no failing builds, so having that be an empty list, rather than an ever-growing list seems sensible.
- The failure mode for not including something in `failing_builds` is lots of noise, which is easy to notice and fix. The failure mode for not including something in `passing_builds` is silent.

**Comment by tomprince** [comment]
This is currently deployed on staging (but the github parts are disabled). I did some manual tests of the github stuff.

File: flocker_bb/zulip_status.py [hunk 2]
```diff
@@ -35,6 +38,12 @@ def __init__(self):
         """
         @param zulip: A zulip client to use to send messages.
         @type zulip: L{_Zulip}
+
+        @param stream: The stream build results should be reported to.
+        @param critical_stream: The stream critical build failures should be
+            reported to.
+        @param failing_builderes: List of builders for which critical failures
```
Comment (adamtheturtle) [inline flocker_bb/zulip_status.py]: failing_builderes > failing_builders

**Pull request merged**